### PR TITLE
Enhance test cases to scale single-node `etcd` to multi-node `etcd` cluster

### PR DESCRIPTION
**Pull request by seshachalam-yv**
**How to categorize this PR?**

/area testing high-availability
/kind enhancement test

**What this PR does / why we need it**:

This PR enhances the existing test cases by adding various scenarios for scaling a single-node `etcd` cluster to a multi-node `etcd` cluster. The new test cases cover the following scenarios:

1. Scaling a single-node `etcd` directly to a multi-node `etcd` cluster.
2. Scaling down a single-node `etcd` to zero nodes and then scaling up to a multi-node `etcd` cluster.

These additional test cases will help ensure the robustness of our `etcd` scaling implementation and increase confidence in the stability of the system.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator

```

### CODE REVIEW

File: test/e2e/etcd_multi_node_test.go [hunk 1]
```diff
@@ -135,29 +135,52 @@ var _ = Describe("Etcd", func() {
...
+			By("Scaling up a healthy cluster (from 1 to 3)")
+			Expect(cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)).To(Succeed())
+			etcd.Spec.Replicas = 3
+			updateAndCheckEtcd(ctx, cl, objLogger, etcd, multiNodeEtcdTimeout)
+
+			By("Deleting the single-node etcd")
+			deleteAndCheckEtcd(ctx, cl, objLogger, etcd, multiNodeEtcdTimeout)
+		})
+
+		It("should scale down a single-node etcd and then scale up to a multi-node etcd cluster", func() {
```
Comment (unmarshall) [inline test/e2e/etcd_multi_node_test.go:158]:
```suggestion
		It("should scale down a single-node etcd to 0 and then scale up to a multi-node etcd cluster", func() {
```

**Comment by unmarshall** [comment]
In the description of this PR you have mentioned:
```
Scaling down a single-node etcd to zero nodes and then scaling up to a multi-node etcd cluster.
Scaling down a single-node etcd to zero nodes, scaling up to one node, and then scaling up to a multi-node etcd cluster.
```
However if one looks at the test then there is just one test -> 1->0 followed by 0->1 followed by 1->3. From that perspective both of the above are doing the same thing. Did you plan to have a test for 1 -> 0, 0 -> 3 ?

**Comment by ishan16696** [comment]
@seshachalam-yv can you rebase this PR on latest master, so that we can see druid e2e tests are working fine or not ?

**Comment by seshachalam-yv** [comment]
> In the description of this PR you have mentioned:
> 
> ```
> Scaling down a single-node etcd to zero nodes and then scaling up to a multi-node etcd cluster.
> Scaling down a single-node etcd to zero nodes, scaling up to one node, and then scaling up to a multi-node etcd cluster.
> ```
> 
> However if one looks at the test then there is just one test -> 1->0 followed by 0->1 followed by 1->3. From that perspective both of the above are doing the same thing. Did you plan to have a test for 1 -> 0, 0 -> 3 ?

Currently we are not supporting above case, so I removed as part of this commit 38f6e09 and updated PR description.

File: test/e2e/etcd_multi_node_test.go [hunk 2]
```diff
@@ -135,29 +135,52 @@ var _ = Describe("Etcd", func() {
...
 			Expect(cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)).To(Succeed())
-			etcd.Spec.Replicas = multiNodeEtcdReplicas
+			etcd.Spec.Replicas = 3
+			updateAndCheckEtcd(ctx, cl, objLogger, etcd, multiNodeEtcdTimeout)
+
+			By("Deleting the single-node etcd")
+			deleteAndCheckEtcd(ctx, cl, objLogger, etcd, multiNodeEtcdTimeout)
+		})
+
+		It("should scale down a single-node etcd to 0, then scale from 0->1 and then from 1->3", func() {
```
Comment (ishan16696) [inline test/e2e/etcd_multi_node_test.go:158]:
```suggestion
		It("should scale down a single-node etcd to 0, then scale up from 0->1 replicas and then from 1->3 replicas", func() {
```

File: test/e2e/etcd_multi_node_test.go [hunk 3]
```diff
@@ -135,29 +135,52 @@ var _ = Describe("Etcd", func() {
...
+			ctx, cancelFunc := context.WithTimeout(parentCtx, 10*time.Minute)
+			defer cancelFunc()
+
+			etcd := getDefaultEtcd(etcdName, namespace, storageContainer, storePrefix, provider)
+			objLogger := logger.WithValues("etcd-multi-node", client.ObjectKeyFromObject(etcd))
+
+			By("Creating a single-node etcd")
+			createAndCheckEtcd(ctx, cl, objLogger, etcd, singleNodeEtcdTimeout)
+
+			By("Scaling down a healthy cluster (from 1 to 0)")
```
Comment (ishan16696) [inline test/e2e/etcd_multi_node_test.go:168]:
```suggestion
			By("Scaling down a healthy cluster (from 1 to 0 replica)")
```

File: test/e2e/etcd_multi_node_test.go [hunk 4]
```diff
@@ -135,29 +135,52 @@ var _ = Describe("Etcd", func() {
...
+			Expect(cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)).To(Succeed())
+			etcd.Spec.Replicas = 0
 			updateAndCheckEtcd(ctx, cl, objLogger, etcd, multiNodeEtcdTimeout)
 
-			// TODO: Uncomment me once scale down replicas from 3 to 1 is supported.
-			// By("Scale down of a healthy cluster (from 3 to 1)")
-			// Expect( cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)).To(Succeed())
-			// etcd.Spec.Replicas = replicas
-			// updateAndCheckEtcd(ctx, cl, objLogger, etcd)
+			By("Scaling up cluster (from 0 to 1)")
```
Comment (ishan16696) [inline test/e2e/etcd_multi_node_test.go:173]:
```suggestion
			By("Scaling up cluster (from 0 to 1 replica)")
```

File: test/e2e/etcd_multi_node_test.go [hunk 5]
```diff
@@ -135,29 +135,52 @@ var _ = Describe("Etcd", func() {
...
+			Expect(cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)).To(Succeed())
+			etcd.Spec.Replicas = 1
+			updateAndCheckEtcd(ctx, cl, objLogger, etcd, multiNodeEtcdTimeout)
+
+			By("Scaling up a healthy cluster (from 1 to 3)")
```
Comment (ishan16696) [inline test/e2e/etcd_multi_node_test.go:178]:
```suggestion
			By("Scaling up a healthy cluster (from 1 to 3 replica)")
```

**Review by ishan16696 — COMMENTED** [review]
few nits

**Review by abdasgupta — APPROVED** [review]
I am fine with the changes. Thanks for the PR. /lgtm

**Review by unmarshall — APPROVED** [review]
/lgtm

**Review by ishan16696 — APPROVED** [review]

**Pull request merged**